modules/core/src/test/java/org/apache/ignite/internal/processors/compute/IgniteComputeWithPendingTopologyDeadlockTest.java

```diff
@@ -0,0 +1,313 @@
+
+    /** Maximum timeout for topology update (prevent hanging and long shutdown). */
+    private static final long PENDING_TIMEOUT = 15_000L;
+
+    /** */
+    private static final long CACHE_CREATION_TIMEOUT = 2_000L;
+
+    /** */
+    private static final int CACHE_SIZE = 10;
+
+    /** */
+    private static final String CACHE1 = "cache1";
+
+    /** */
+    private static final String CACHE2 = "cache2";
+
+    /** */
+    private static final String CACHE3 = "cache3";
+
+    /** */
+    private static final String EXPECTED_ERROR_MESSAGE = "Pending " +
+        "topology found - job execution within transaction or lock was canceled.";
+
+    /** {@inheritDoc} */
+    @Override public void beforeTest() throws Exception {
+        super.beforeTest();
+        startGrid(0);
+    }
+
+    /** {@inheritDoc} */
+    @Override public void afterTest() throws Exception {
+        super.afterTest();
+        stopAllGrids();
+    }
+
+    /**
+     * Create cache configuration with transactional atomicity mode.
+     *
+     * @param name Cache name.
+     * @param mode {@link CacheMode Cache mode.}
+     * @param writeMode {@link CacheWriteSynchronizationMode Cache write synchronization mode.}
+     * @return Cache configuration.
+     */
+    private CacheConfiguration<Integer, Integer> cacheConfiguration(String name, CacheMode mode,
+        CacheWriteSynchronizationMode writeMode) {
+        CacheConfiguration<Integer, Integer> cacheConfiguration = new CacheConfiguration<>(name);
+
+        cacheConfiguration.setCacheMode(mode);
+        cacheConfiguration.setWriteSynchronizationMode(writeMode);
+        cacheConfiguration.setAtomicityMode(CacheAtomicityMode.TRANSACTIONAL);
+
+        return cacheConfiguration;
+    }
+
+    /**
+     * Explicit lock test.
+     *
+     * @throws Exception If fails.
+     */
+    public void testExplicitLock() throws Exception {
+        for (final CacheMode mode : CacheMode.values()) {
+            for (final CacheWriteSynchronizationMode syncMode : CacheWriteSynchronizationMode.values()) {
+                if (mode == CacheMode.LOCAL) {
+                    // Expecting normal execution of distributed task.
+                    doTestExplicitLock(mode, syncMode);
+                }
+                else {
+                    // Expecting that execution will be rejected.
+                    assertThrows(log, new Callable<Void>() {
+                        @Override public Void call() throws Exception {
+                            doTestExplicitLock(mode, syncMode);
+
+                            return null;
+                        }
+                    }, IgniteCheckedException.class, EXPECTED_ERROR_MESSAGE);
+                }
+            }
+        }
+    }
+
+
+    /**
+     * Transactional lock test.
+     *
+     * @throws Exception If fails.
+     */
+    public void testTransactionLock() throws Exception {
+        for (final CacheMode cacheMode : CacheMode.values()) {

```

Comment:
Please, make 3 tests instead this for loop

`testTransactionLockLocal`
`testTransactionLockReplicated`
`testTransactionLockPartitioned`

In case of any future bugs, it speeds up understanding and fixing.